apps/dotcom/sync-worker/src/TLAppDurableObject.ts

```diff
@@ -0,0 +1,269 @@
+		}
+	}
+
+	// Handle a request to the Durable Object.
+	async fetch(req: IRequest) {
+		try {
+			return await this.router.fetch(req)
+		} catch (err) {
+			console.error(err)
+			return new Response('Something went wrong', {
+				status: 500,
+				statusText: 'Internal Server Error',
+			})
+		}
+	}
+
+	async onRequest(req: IRequest) {
+		// extract query params from request, should include instanceId
+		const url = new URL(req.url)
+		const params = Object.fromEntries(url.searchParams.entries())
+		let { sessionId } = params
+
+		// handle legacy param names
+		sessionId ??= params.sessionKey ?? params.instanceId
+
+		// Create the websocket pair for the client
+		const { 0: clientWebSocket, 1: serverWebSocket } = new WebSocketPair()
+		serverWebSocket.accept()
+
+		const room = await this.getRoom()
+
+		// all good
+		room.handleSocketConnect({
+			sessionId: sessionId,
+			socket: serverWebSocket,
+		})
+		return new Response(null, { status: 101, webSocket: clientWebSocket })
+	}
+
+	triggerPersistSchedule = throttle(() => {
+		this.persistToDatabase()
+	}, 2000)
+
+	// Load the room's drawing data. First we check the R2 bucket, then we fallback to supabase (legacy).
+	async loadFromDatabase(): Promise<DBLoadResult> {
+		try {
+			const [
+				{
+					results: [topic],
+				},
+				{ results: records },
+			] = await this.env.DB.batch([
+				this.loadTopic.bind(this.userId),
+				this.loadRecords.bind(this.userId),
+			])
+			const snapshot: RoomSnapshot = {
+				clock: 0,
+				documents: records.map(({ record, lastModifiedEpoch }: any) => ({
+					state: JSON.parse(record),
+					lastChangedClock: lastModifiedEpoch,
+				})),
+				schema: JSON.parse((topic as any)?.schema ?? null) ?? tldrawAppSchema.serialize(),
+				tombstones: JSON.parse((topic as any)?.tombstones ?? null) ?? {},
+			}
+			// when loading, prefer to fetch documents from the bucket
+			return { type: 'snapshot_found', snapshot }
+		} catch (error) {
+			console.error('failed to fetch doc', error)
+			return { type: 'error', error: error as Error }
+		}
+	}
+
+	_lastPersistedClock: number | null = null
+	_persistQueue = createPersistQueue(async () => {
+		// check whether the worker was woken up to persist after having gone to sleep
+		if (!this._room) return
+		const room = await this.getRoom()
+		const clock = room.getCurrentDocumentClock()
+		if (this._lastPersistedClock === clock) return
+		const snapshot = room.getCurrentSnapshot()
+
+		let deletedIds = [] as string[]
+		let updatedRecords = [] as { state: any; lastChangedClock: number }[]
+		if (typeof this._lastPersistedClock === 'number') {
+			deletedIds = Object.entries(snapshot.tombstones ?? {})
+				.filter(([_id, clock]) => clock > this._lastPersistedClock!)
+				.map(([id]) => id)
+
+			updatedRecords = snapshot.documents.filter(
+				(doc) => doc.lastChangedClock > this._lastPersistedClock!
+			)
+		}
+
+		await this.env.DB.batch([
+			this.updateTopic.bind(
+				this.userId,
```

Comment:
these kinds of bindings make me nervous about keeping things in sync with the ordering. is there not a named binding possibility? 
i'm thinking of something like `queryRaw` in `prisma` that at least let's you do JS templating: https://www.prisma.io/docs/orm/prisma-client/using-raw-sql/raw-queries

Reply:
the stock D1 prepared statement api doesn't support named params yet.